### ISSUE TITLE
Move ResourceID to fabric8 kubernetes client

### ISSUE THREAD

**Issue by lburgazzoli** [java-operator-sdk/java-operator-sdk#1067]
The `io.javaoperatorsdk.operator.processing.event.ResourceID` class seems to be useful also outside the operator, it would be nice to move it to the fabric8 kubernetes client

**Comment by metacosm** [comment]
I think we had discussed this at some point but I guess that didn't go further :)
It would maybe make more sense to open that issue in the fabric8 client project, though?

**Comment by lburgazzoli** [comment]
I will tomorrow

**Comment by lburgazzoli** [comment]
https://github.com/fabric8io/kubernetes-client/issues/3991

**Comment by csviri** [comment]
To mention also here, this is called `ObjectKey`, in controller runtime.
https://github.com/kubernetes-sigs/controller-runtime/blob/c46b410039fdc9d620a4217bc7c63c5a7fffae46/pkg/client/interfaces.go#L31-L31

Would be a little better to have consistent naming, but it's not that important.

**Comment by metacosm** [comment]
Maybe worth renaming it before we release 3.0?

**Comment by metacosm** [comment]
Actually looking at the code it's actually named `NamespacedName` and it's coming from `apimachinery`… I'd prefer using `NamespacedName` over `ObjectKey` actually :)

**Comment by csviri** [comment]
created this: https://github.com/java-operator-sdk/java-operator-sdk/issues/1073

I like ObjectKey more, there is a reason why it's probably also aliased in controller runtime. It describes better to identify an Object.

**Comment by csviri** [comment]
To be consistent actually fabri8 client should have a class with NamespacedName and we would just to extend it.

**Comment by metacosm** [comment]
> created this: #1073
> 
> I like ObjectKey more, there is a reason why it's probably also aliased in controller runtime. It describes better that is used to identify an Object IMHO.

I don't agree. It's aliased in controller runtime because it's used in a cache-like interface. That's not strictly what we're using it for and `NamespacedName` corresponds much better to the intent as opposed to something that's both too generic (what is the `Object` in `ObjectKey` referring to?) and too restrictive (the `Key` part implies it's used only in map-like structures)… I'd rather we keep `ResourceID` than use `ObjectKey` if we don't want to use `NamespacedName`.

**Comment by csviri** [comment]
> I don't agree. It's aliased in controller runtime because it's used in a cache-like interface.

It's used to identify an object/resource [specify the object](http://github.com/kubernetes-sigs/controller-runtime/blob/c46b410039fdc9d620a4217bc7c63c5a7fffae46/pkg/client/metadata_client.go#L119-L119 ) of a given type.
(The client here does not necessary use cache.) This is really just the information needed to identify the object in case the **other  required information is present**.  Mainly because of this I don't like the `ID` in the name. This is not really an id, it does not identify the resource without additional information (api group, kind, version). So that's why I think `Key` is better.

Object and Resource are basically synonyms, so if we change it, would be either `ResourceKey` or `ObjectKey`. I'm fine with either. Since this is how we use it. 
See the more correct definition between object and resource in kubernetes: https://stackoverflow.com/questions/52309496/difference-between-kubernetes-objects-and-resources   - the Object seems to be the correct term here to me.

Also this is quite obvious it's a Kubernetes object from the context, not a Java Object. We already use `ObjectMeta`, `ObjectList` and other names like this. Actually it's better in this sense, that in fabric8 client this is used. 

So that is why I think that using `Object` and `Key` is the good combination here. But if you think `ResourceKey` is better, I'm fine with that too. 

If there is no consensus, we can just keep it as it is.

**Comment by metacosm** [comment]
The problem is that `Resource` has a specific meaning for us which doesn't really map to the Kubernetes concept of resource. So any attempt to use something from controller-runtime will just be confusing in my opinion. For example, `ResourceID` is used in JOSDK as key for external objects as well. Trying to use the name from controller-runtime when it's actually used for something different would be a mistake, and actually, this might also mean that porting the concept to the fabric8 client might not be such a good idea, unless we clarify the semantics (and in this case, the concept of `NamespacedName` is much more accurate than `ObjectKey` for how we use it, imo).

**Comment by csviri** [comment]
>  For example, ResourceID is used in JOSDK as key for external objects as well

Where exactly? I don't think so. It always targets a Kubernetes Object.

>  Trying to use the name from controller-runtime when it's actually used for something different would be a mistake

It's not really something different to my understanding, they are also just addressing an object. 
But that is why I tried to describe the rational behind the name above. (So why `Key` instead of `ID`, and why `Object` instead of `Resource`) 

Why I don't like NamespacedName is because it's a very generic concept. It really not intention revealing, we want to identify an object by the `ResourceID`/`ObjectKey`. (It' not fundamntelly wrong by no means, it just does not feel very cohesive, or on the same level of abstraction as the actual context) 

> this might also mean that porting the concept to the fabric8 client might not be such a good idea, unless we clarify the semantics 

I don't have strong opinion on this. Maybe we should involve @manusa and @shawkins for that.

**Comment by metacosm** [comment]
> > For example, ResourceID is used in JOSDK as key for external objects as well
> 
> Where exactly? I don't think so. It always targets a Kubernetes Object.

`CachingEventSource` for example. The target type is not necessarily a Kubernetes object.

> 
> > Trying to use the name from controller-runtime when it's actually used for something different would be a mistake
> 
> It's not really something different to my understanding, they are also just addressing an object. But that is why I tried to describe the rational behind the name above. (So why `Key` instead of `ID`, and why `Object` instead of `Resource`)

Well, for JOSDK, a `Resource` is anything that needs to be managed, it's not restricted to Kubernetes objects for sure. So in JOSDK a `Resource` might be a Kubernetes Object but it might be a MySQL database or an external API.

> Why I don't like NamespacedName is because it's a very generic concept. It really not intention revealing, we want to identify an object by the `ResourceID`/`ObjectKey`. (It' not fundamntelly wrong by no means, it just does not feel very cohesive, or on the same level of abstraction as the actual context)

I agree that `NamespacedName` is very generic but so is `ObjectKey`. `ObjectKey` is worse for us because we have nothing that we call `Object` in our code, so it doesn't refer to any concept we already use. Probably `NamespacedName` is not good either, so let's keep `ResourceID` :)

**Comment by csviri** [comment]
Ok, I can live with that :)
